### PR TITLE
Add alignment and start/end angle to pie plots

### PR DESCRIPTION
**Pull request by carlodx**
Addresses issue [#1041](https://github.com/palantir/plottable/issues/1041)

Adding to pie plot: 
- `startAngle(angle: number)`:  sets the start angle, so pies do not need to be a complete circle
- `endAngle(angle: number)` : sets the end angle, so pies do not need to be a complete circle
- `hAlign(alignment: string)`: the alignment is by default in the center of the render area. using `"left"` or `"right"` you can align the horizontal axis to the left or the right
- `vAlign(alignment: string)`: the alignment is by default in the center of the render area. using `"top"` or `"bottom"` you can align the vertical axis to the top or the bottom.

### CODE REVIEW

**Comment by palantirtech** [comment]
Thanks for your interest in [palantir/plottable](https://github.com/palantir/plottable), @carlosdiazh! Before we can accept your pull request, you need to sign our contributor license agreement - just visit https://cla.palantir.com/ and follow the instructions. Once you sign, I'll automatically update this pull request.

File: test/plots/piePlotTests.ts [hunk 1]
```diff
@@ -142,6 +142,140 @@ describe("Plots", () => {
...
+        assert.strictEqual(piePlot.startAngle(expectedStartAngle), piePlot, "setter returns the calling pie plot");
+        let arcPaths = piePlot.content().selectAll(".arc.fill");
+        assert.strictEqual(arcPaths.size(), 2, "has two sectors");
+
+        let pathString0 = d3.select(arcPaths[0][0]).attr("d");
+        let decomposedPath0 = TestMethods.decomposePath(TestMethods.normalizePath(pathString0));
+
+        let moveCommand0 = decomposedPath0[0];
+        assert.closeTo(moveCommand0.arguments[0], 0, 1, "starts first arc at the bottom of the circle (x)");
+        assert.operator(moveCommand0.arguments[1], ">", 0, "starts first arc at the bottom of circle (y)");
```
Comment (hellochar) [inline test/plots/piePlotTests.ts]: nit: I find assert.isAbove and assert.isBelow slightly more readable

File: src/plots/piePlot.ts [hunk 2]
```diff
@@ -198,6 +204,53 @@ namespace Plottable.Plots {
...
+      } else {
+        this._startAngle = angle;
+        this._updatePieAngles();
+        this.render();
+        return this;
+      }
+    }
+
+    /**
+     * Gets whether slice labels are enabled.
```
Comment (hellochar) [inline src/plots/piePlot.ts]: update docs here

File: src/plots/piePlot.ts [hunk 3]
```diff
@@ -276,8 +375,9 @@ namespace Plottable.Plots {
       let dataset = this.datasets()[0];
       let data = this._getDataToDraw().get(dataset);
       let pie = d3.layout.pie().sort(null).value((d, i) => sectorValueAccessor(d, i, dataset))(data);
-      this._startAngles = pie.map((slice) => slice.startAngle);
-      this._endAngles = pie.map((slice) => slice.endAngle);
+      let factor = (this.endAngle() - this.startAngle()) / (2 * Math.PI);
```
Comment (hellochar) [inline src/plots/piePlot.ts]: Looks like d3.layout.pie has something similar already: see https://github.com/d3/d3-3.x-api-reference/blob/master/Pie-Layout.md#startAngle . Can we take advantage of that?

**Review by hellochar — CHANGES_REQUESTED** [review]
Thanks for the PR @carlosdiazh ! 

Code looks pretty straightforward, but I'm not sure I understand the purpose of hAlign/vAlign. Pie plots don't have vertical or horizontal axes. The code looks like it just shifts the entire pie within the rendering area?

**Comment by carlodx** [comment]
@hellochar The purpose of `hAlign` and `vAlign` is to be able to place the center of the pie in one of the sides of the render area, so that when you are drawing a half pie, half of the render area isn't left out blank. This is the only thing I could think of, but if there is something else I could use I'd be happy to change it.

Also, I've addressed the rest of the change requests

**Comment by hellochar** [comment]
Ah I see - so you mean when setting startAngle/endAngle by itself, you'll get pies that only cover half the area: 
image

In order to fix that, users can do `vAlign("top")` to push it to the top: 
image

But won't that have the same issue where the second half of the chart is completely empty? Furthermore, you never want to use `hAlign/vAlign` by itself (otherwise the pie would escape its drawing boundary); you only want to set them to certain values in combination with `startAngle`/`endAngle`, is that right? I'm worried that exposing this API surface area will confuse users since it's not obvious what it does/what it's used for. 

How would you feel about auto-scaling the pie chart to fill the plot width/height depending on the start/endAngle?

@CalvinFernandez @themadcreator thoughts?

File: src/plots/piePlot.ts [hunk 4]
```diff
@@ -198,6 +204,53 @@ namespace Plottable.Plots {
...
       return this;
     }
+    /**
+     * Gets the start angle of the Pie Plot
+     *
+     * @returns {number} Returns the start angle
+     */
+    public startAngle(): number;
+    /**
+     * Sets the end angle of the Pie Plot.
```
Comment (CalvinFernandez) [inline src/plots/piePlot.ts]: Sets the ~end~ start angle of the Pie Plot

File: src/plots/piePlot.ts [hunk 5]
```diff
@@ -198,6 +204,53 @@ namespace Plottable.Plots {
...
+     *
+     * @param {number} endAngle
```
Comment (CalvinFernandez) [inline src/plots/piePlot.ts]: ~endAngle~ start angle

File: src/plots/piePlot.ts [hunk 6]
```diff
@@ -222,6 +275,52 @@ namespace Plottable.Plots {
       }
     }
 
+   /**
+     * Get whether slice labels are enabled.
+     *
+     * @returns {boolean} Whether slices should display labels or not.
```
Comment (CalvinFernandez) [inline src/plots/piePlot.ts]: string

**Review by CalvinFernandez — COMMENTED** [review]
This is great! Thanks for the PR @carlosdiazh! re: @hellochar's comment, I think one solution could be to adjust the size of the component during the space negotiation phase to ensure that the chart always correctly fills the plot area. Also not a huge fan of `hAlign/vAlign` .  I'm also fine with just leaving the second half of the chart completely empty and revisiting this as a follow up commit.

File: src/plots/piePlot.ts [hunk 7]
```diff
@@ -30,8 +32,15 @@ namespace Plottable.Plots {
 
     public computeLayout(origin?: Point, availableWidth?: number, availableHeight?: number) {
       super.computeLayout(origin, availableWidth, availableHeight);
-      this._renderArea.attr("transform", "translate(" + this.width() / 2 + "," + this.height() / 2 + ")");
-      let radiusLimit = Math.min(this.width(), this.height()) / 2;
+      
+      let pieCenter = this._pieCenter()
+      this._renderArea.attr("transform", "translate(" + pieCenter.x + "," + pieCenter.y + ")");
```
Comment (hellochar) [inline src/plots/piePlot.ts:40]: We should update the transform in the minAngle/maxAngle setter, not just in `computeLayout`

File: src/plots/piePlot.ts [hunk 8]
```diff
@@ -275,11 +331,54 @@ namespace Plottable.Plots {
...
       let data = this._getDataToDraw().get(dataset);
-      let pie = d3.layout.pie().sort(null).value((d, i) => sectorValueAccessor(d, i, dataset))(data);
+      let pie = d3.layout.pie().sort(null).startAngle(this._startAngle).endAngle(this._endAngle)
+        .value((d, i) => sectorValueAccessor(d, i, dataset))(data);
       this._startAngles = pie.map((slice) => slice.startAngle);
       this._endAngles = pie.map((slice) => slice.endAngle);
     }
 
+    private _pieCenter(): Point {
+      let a = this._startAngle < this._endAngle ? this._startAngle: this._endAngle;
```
Comment (hellochar) [inline src/plots/piePlot.ts:341]: Holy cow this is impressive. Can we add comments to describe what's going on? Is there a way to reduce the number of if/elses?

**Review by hellochar — CHANGES_REQUESTED** [review]
Thanks for removing hAlign/vAlign. Feels close now; just some polish on the new pie centering logic

File: src/plots/piePlot.ts (hunk 8 above)
Comment (carlodx) [reply src/plots/piePlot.ts:341]: Sure, I'll add some comments. 
I've been trying to come up with a more generic way to compute the center but I haven't been able to find anything that doesn't leave out some possibilities potentially cutting out part of the pie.

File: src/plots/piePlot.ts (hunk 7 above)
Comment (carlodx) [reply src/plots/piePlot.ts:40]: I tried doing this, but `this._renderArea` is still undefined

File: src/plots/piePlot.ts (hunk 8 above)
Comment (carlodx) [reply src/plots/piePlot.ts:341]: I've added a general comment, above the if/elses, it's a bit hard to explain exactly whats going on on each condition without drawing. Hopefully this'll do. Let me know though. 

We could also get rid of the last if on each block but I feel like stating the condition makes it easier to understand so one can picture in their head the image of the area of the whole circle the pie should take.

**Review by hellochar — APPROVED** [review]

**Comment by hellochar** [comment]
Looks great, thanks @carlosdiazh !

**Comment by hellochar** [comment]
Oops, I let https://github.com/palantir/plottable/issues/3168 through... @carlosdiazh if you're hitting that feel free to submit another PR to fix. If not, us Plottable folks will triage and deal with it eventually

**Pull request merged**